### PR TITLE
feat: 1534 service window in summary report

### LINKED ISSUE #1534: feat: Include service window in Summary report
### Describe the problem

Currently, the summary section of the validation report shares the `feed_start_date` and `feed_end_date` when it's provided. However, many feeds don't include `feed_info.txt`, which makes it difficult to know the service window at a glance.

### Proposed solution

- Add full service date range to summary report. [Calculation here](https://github.com/MobilityData/gtfs-validator/blob/master/main/src/main/java/org/mobilitydata/gtfsvalidator/util/CalendarUtil.java)

- Pseudologic for the date range here: 

- When only `calendars.txt` is used: Loop through all services referenced with a trip_id in `trips.txt`. Loop through all those services with a trip_id in`calendars.txt`. check that there have an `start_date` and `end_date` - take the earliest start date and the latest end date
When only `calendar_dates.txt` is used: Loop through all services referenced with a trip_id in `trips.txt`. Loop through all those services in `calendar_dates.txt`. Take the earliest date and the latest date
When both are used: Loop through all services referenced with a trip_id in `trips.txt`. Calculate the earliest start date and the latest end date for each service with a trip_id using `calendars.txt` , and then check to make sure there isn't a service exception in `calendar_dates.txt` on either of those dates

### Design considerations

- The summary report is getting busy [(example here)](https://gtfs-validator-results.mobilitydata.org/50456ba1-92f2-45c5-a6f8-524087834eac/report.html). How do we make sure this info fits and is intuitive?

### Alternatives you've considered

_No response_

### Additional context

- Include both feed date range and service range
- In the future, calculate based on majority of dates
- Look at other validators' approach to calendar dates, e.g Google Partner Dashboard
_No response_

### PR DESCRIPTION
**Pull request by qcdyx**
**Summary:**
Closes #1534 

**Expected behavior:** 
GTFS Validator report should show Service Date Range under Feed Info section.

1. When only calendar_dates.txt is used 
test dataset: 
[poa_gtfs.zip](https://github.com/user-attachments/files/16949788/poa_gtfs.zip)
) 
image

2.  When only calendar_dates.txt is used 
test dataset: 
[Google_transit.zip](https://github.com/user-attachments/files/16949823/Google_transit.zip)
image

3. When both calendars.txt and calendar_dates.txt are used 
test dataset: 
[Archive.zip](https://github.com/user-attachments/files/16949852/Archive.zip)
image

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [ ] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:81]: This requires that all 4 files be present, but calendar and calendar_date are conditionally required, meaning one of the two could be absent.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:410]: Are you sure tripContainer cannot be null?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:431]: Can these be null or will they be some default (e.g. 1970-01-01) if the date is not in the file?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:449]: Maybe you should check if the exception_type is addition before using it in the calculation.

File: main/src/test/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadataTest.java
Comment (jcpitre) [inline main/src/test/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadataTest.java:39]: typo

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:431]: you are tight, they can be unix time 1970-01-01, in that case, we should just display N/A

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:449]: When only calendar_dates.txt is used, can we assume that the exception_type is always 1? @emmambd

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:417]: Well, if trips.txt is absent it will give a error notice, but do we have a guarantee that the code here will not be reached in that case? If it is, we'll get an NPE which in general is not good.

Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:417]: Ok, it could be empty. I'll add the logic.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:495]: Won't this generate an NPE if earliestStartDate is null and latestEndDate is not null?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:426]: Can't this loop be replaced by something like this: 
GtfsCalendar calendar = ((GtfsCalendarTableContainer) calendarTable).byServiceId(serviceId);

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:447]: Can't this loop be replaced by something like this:
GtfsCalendarDate calendarDate = ((GtfsCalendarDateTableContainer) calendarDateTable).byServiceId(serviceId);

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:380]: With the change done earlier tripContainer, calendarTable and calendarDateTable are not used any more within this method, so they should be removed.

Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:380]: removed!

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:418]: Yes, it could, but I prefer to keep the loop as is for the moment. We can create new issue that optimize the code using Optional handling.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:442]: Same here.

**Comment by emmambd** [comment]
Hey @qcdyx - I'm able to do the spot checks now to see if we need service date ranges in cases where feed_info start date and end date apply! 

However, I'm still seeing the bug that was fixed in https://github.com/MobilityData/gtfs-validator/pull/1798. Example report here: 

[report.json](https://github.com/user-attachments/files/17104266/report.json)

Since this was merged to master, why would this be happening?

**Review by emmambd — CHANGES_REQUESTED** [review]
Great work @qcdyx! After doing some spot checks, it makes sense for us to include both the feed start and end date values and the service date range values. 

Minor changes: 

• Change "Service Date Range" to "Service Dates"
• Add tooltip to Service Date Range on the HTML report: 

"The range of service dates covered by the feed, based on trips with an associated service_id in calendar.txt and/or calendar_dates.txt."

**Comment by emmambd** [comment]
@qcdyx Sorry Jingsi, one more minor requirements change and then you can merge 🙈 In thinking about how we'll eventually add majority service window to this report, let's change "service dates" to "service window" so we'll have consistency between these concepts. 

Once changed, all good and ready to merge :)

**Comment by davidgamez** [comment]
@qcdyx  The acceptance tests are reporting a few(20) feeds sources `corrupted`. This means the report was not generated for some reason, and the comparison script cannot compare the previous/latest reports. One of the possible reasons is that the validator throws a critical exception while processing the feed. I tried the `ca-ontario-guelph-transit-gtfs-1300` locally with this branch and indeed raised a NullPointerException in [line](https://github.com/MobilityData/gtfs-validator/blob/9d12dc9f3414c5373e585e164282c4084c91b81b/main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java#L481). In this case the service Id was not found in the `servicePeriods` Map. We can target this specific NullPointerException. 
In addition, I'm suggesting to wrap with a try/catch the full `"discovery"` of the service date between line [429](https://github.com/MobilityData/gtfs-validator/blob/9d12dc9f3414c5373e585e164282c4084c91b81b/main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java#L429) to [507](https://github.com/MobilityData/gtfs-validator/blob/9d12dc9f3414c5373e585e164282c4084c91b81b/main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java#L507), this will make sure that any exception raised due to a null object is properly caught and the report is generated. I am open to suggestions on how to make sure that the report is generated. 

After the changes, let's examine the corrupted files in the acceptance test and ensure that the validator generates the report.

**Comment by davidgamez** [comment]
This looks good for the scope of the issue. I did create a follow-up issue #1848 to address the pending corrupted feeds.

**Review by davidgamez — APPROVED** [review]
LGTM

**Comment by cka-y** [comment]
Was the service window added to the json report? It's part of the tasking, but couldn't find the information from running the jar from this branch

**Comment by qcdyx** [comment]
Correct me if I'm wrong but I remembered that we talked about splitting the ticket and focusing on just the service date range on UI when we did the estimate. I recommend merging this PR now, and I’ll create a separate one to add it to the JSON report coz I need to merge master changes into my branch and it takes some time to go thru all the checks. @emmambd @davidgamez

**Comment by emmambd** [comment]
I'm fine with splitting, but the next priority issue will be adding it to the JSON report in that case, since we can't make any changes to the Mobility Database we want without the JSON report functionality. cc @davidgamez for final call

**Comment by qcdyx** [comment]
@emmambd could you approve my PR since merging is blocked due to requested changes.

**Comment by emmambd** [comment]
Did you and @davidgamez discuss? Just waiting for his thoughts on splitting or not before I approve it.

**Comment by davidgamez** [comment]
> Did you and @davidgamez discuss? Just waiting for his thoughts on splitting or not before I approve it.

Yes the JSON changes can be part of a different issue.

**Review by emmambd — APPROVED** [review]

**Pull request merged**
